### ISSUE TITLE
[WIP] 💬 RFC: Dynamic microservice plugins for front-end

### ISSUE THREAD

**Issue by tumido** [backstage/backstage#17054]
### 🔖 Need

We want to be able to manage a plugin lifecycle separately from the main backstage core application. This includes installing new plugins and removing and upgrading them. Plugins should be able to communicate with other Backstage APIs and services while being individually scalable, deployable, and pluggable through configuration. Adding or removing a plugin from Backstage should not require NPM package installation and Backstage image rebuilds and redeployments. Instead, we want to be able to add plugins as dynamic content. 

This RFC is limited to front-end plugins only. Frontend plugins should not be restricted to specific packaging and deployment types but should be mountable from any HTTP service provider or CDN directly.

### 🎉 Proposal

We would like to use the Webpack module federation for most of the heavy lifting. In Red Hat, we’ve successfully used this approach for a multitude of products and we would like to donate this user experience to Backstage as well. Red Hat is currently in process of decoupling this federation codebase into a standalone pluggable framework:

https://github.com/scalprum/scaffloding

This framework has minimal dependencies and serves as a wrapper on top of the Webpack module federation providing a unified API.

#### Plugin side changes

We don’t want users to do extensive plugin rewrites or major disruptive code changes. We also want all the plugins to be backward compatible. Currently, plugins are built by Rollup. We propose to extend the build pipeline in `@backstage/cli` with an additional build task, that would invoke Webpack instead (in order to use Webpack module federation, we need to use Webpack instead of Rollup). This additional build task is used only in case the plugin is compatible with this RFC and produces an additional set of assets. As a result, the plugins feature 2 sets of assets, one for the standard plugin build via Rollup (in `dist` folder) and one extra for Webpack module federation (in `dist-scalprum`). If the additional set of assets will be built is decided based on the plugin’s `package.json` content via the addition of:

```
  "scalprum": {
    # Webpack module name, same as used later in the Scalprum registry
    "name": "backstage.plugin-github-actions",
    # Modules exposed through Webpack module federation
    "exposedModules": {
      "./EntityGithubActionsContent": "./src/EntityGithubActionsContent",
      "./EntityRecentGithubActionsRunsCard": "./src/EntityRecentGithubActionsRunsCard"
    }
  },
```

Once built, this additional set of assets contains built JS bundle + a `plugin-manifest.json` for Scalprum.

#### Backstage instance changes

We propose that the plugin discovery should be handled statically through Backstage configuration file:

```
# app-config.yaml
app:
  dynamicPlugins:
    # list all remote plugins
    registry:
      backstage.plugin-entity-feedback:
        # manifest location can refer to either an absolute location or relative location proxied through backstage proxy 
        manifestLocation: '/api/proxy/scalprum/backstage.plugin-entity-feedback/plugin-manifest.json'
        # manifestLocation: 'http://localhost:8003/plugin-manifest.json'

proxy:
  '/scalprum/backstage.plugin-entity-feedback':
    target: 'http://127.0.0.1:8003'
    changeOrigin: true
    secure: false

```

Caveat: Webpack requires packages to define `basePath`. It can be an absolute path containing the host or it can be a relative path. While the relative paths provide us with more flexibility (the package can be built once and hosted multiple times), it poses a challenge for module federation. When module federation is used, Webpack naturally fetches resources with the relative path from the same host as the rest of the bundle. In the case of the local setup of Backstage this defaults to `http://localhost:3000/`. This creates a problem for our proxied approach - backstage needs to proxy requests from frontend to the backend (since backstage proxy is served on the backend). This can be achieved by extending the `package.json` in `packages/app` with WebpackDevServer proxy:

```
  "proxy": {
    "/api": {
      "target": "http://localhost:7007"
    }
```

A remote package declared in the app-config `app.dynamicPlugins.registry` is consumed by Backstage’s `AppManager` and passed to `ScalprumContext`. This makes the remote plugin globally available and consumable from within Backstage front-end.

Simple remote components can be used via:

```
<ScalprumComponent
  scope="backstage.plugin-github-actions"
  module="./EntityGithubActionsContent"
/>
```

While this component is loaded from a remote location, this approach still requires manual code changes to mount it in DOM. That is not desirable, the end goal is to do so through configuration changes. This can be achieved by a set of helper components and config values that define user-configurable mount points:

(This is WIP):
```
app:
  dynamicPlugins:
    mounts:
      # mount point name referenced in a helper React component which pulls in all relevant remote plugins
      catalog.entity-page.tab:
          # name from the plugin registry
        - name: 'backstage.plugin-github-actions'
          module: './EntityGithubActionsContent'
          # additional entity filter
          filter:
            kind: Component
            spec.type: service
          # subroute
          path: /ci-cd-external
          # additional props
          title: 'CI/CD (Scalprum)'
```

These endpoints are used by the Backstage instance frontend code and instance maintainers decide where and how to use these mount points. This addresses the need to support diverse Backstage instances, where each instance can be totally different from the other ones. To explain this in simple terms:

- **WHAT** is mounted is defined in `app-config.yaml` (therefore extensible without code changes)
- **WHERE** are the components mounted is decided by the Backstage instance maintainer through code changes.

WIP example:

```
# EntityLayout.tsc

<EntityLayoutWrapper>
  <DymanicPluginsMount mount="catalog.entity-page.tab">
</EntityLayout.Route>
```

#### Common changes

Since Backstage builder/bundler is defined in `@backstage/cli` we need to adjust this builder. This includes sharing/adding all the shared/singleton packages that need to be shared between the Backstage frontend (host) and the external plugin - packages like `react`, `react-dom`, etc. This may pose a risk that if a plugin requires a singleton package on host, this needs to be adjusted on both sides - in the host app build as well as in the plugin build.

### 〽️ Alternatives

TBD

### ❌ Risks

This introduces changes to the way how `packages/app` and individual plugins are being built. Changes are minimal (adding one Webpack plugin to the `packages/app`) on the host side. On the other hand, on the plugin side, we're adding a brand-new build strategy that requires to be supported. However, it is still an additional strategy - it preserves the current package build, therefore current packages should not be affected.

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by mario-mui** [comment]
https://github.com/module-federation/universe/tree/main/packages/utilities why not use this instead of https://github.com/scalprum/scaffolding

maybe we just need to settle getting manifest info ?

**Comment by Rugvip** [comment]
Thank you for this proposal and pushing this conversation forward! 🎉 

Couple of general comments:

I think the **need** section can be fleshed out more with some argument towards why this would be beneficial. Right now I feel it basically just claims that we need what's proposed in the RFC, without laying the foundation of why. The RFC does propose a more specific implementation, but taking a step back I feel it would be useful to argue the more general case. This is important when it comes to prioritizing the OSS roadmap and where we should be focusing.

Similarly I think it's important to carefully consider alternatives too, so filling out the **alternatives** section would also move this forward.

On the proposal:

1. I really want to not lock us into Webpack, it's slow and in many cases unnecessarily complex for what we're doing. Ideally I would like it to be feasible to move over to a frontend build system where each plugin is built in isolation, so think ~100 parallel builds without excessive build times. The idea there is that backend plugins would take part in the module federation and host their own frontend code, leading to just a single deployable unit rather than the frontend/backend split that we have today. It's all further future ofc and it could be that Webpack is able to handle that setup too, but I would prefer to not couple ourselves too tight to Webpack to keep options open. How's the performance of your current setup, and how many separate compilation units? Any particular tricks to make Webpack perform well?

1. I do have a question on how you envision the distribution of the federated modules to work? as I don't feel that's quite captured here. Is the idea that the packages that are published to NPM include the full federated modules build? or is that build just something that happens locally, and you're then responsible for setting up the rest of the pipeline yourself? I guess what I'm looking for here is a bit more of an end-to-end solution for how we could ship this in OSS, even though we ofc leave space for bespoke solutions too.

1. Related to the above, I do feel there's an opportunity to provide a much neater user experience than manually listing manifest endpoints. For example something like a `manifest-manager` plugin that is able to pull in manifests from various locations, such as external buckets, endpoints, or service discovery. In general I feel that the manifest management is something that's best left as a pluggable API where we can build out solutions in OSS as well as adding additional internal discovery mechanisms.

1. As mentioned in the Adoption SIG WG, there's quite a lot to be figured out in the plugin wiring. Do you feel that there is some problems/solutions in the plugin installation area that we could split out and handle separately?

**Comment by ericis** [comment]
I was a little thrown off by the title, since it states a very specific use of "dynamic Microservices plugins" rather than the capability of something like "Front-end as a Service". I was also confused about Front-end as a Service vs. the Webpack Micro Front-end (MFE) hosts and remotes; for example, it's possible to implement the MFE pattern on a Content Delivery Network (CDN) without any APIs / services.

I would personally like to see the Backstage front-end plug-in architecture to consider the MFE pattern and any opportunity to partially or completely replace the engine behind plugins with MFE. I should note that it is entirely possible to implement MFE in a custom instance of Backstage today with some ingenuity, but would personally like to see it as a foundational capability out-of-the-box.

For me, if MFE were supported out-of-the-box, then front-end as a service is more of an implementation concern for remote endpoints and less of an architectural decision for Backstage to make unless there are specific capabilities the system could provide "on top of" the MFE pattern that necessitate web services instead of a simple CDN endpoint to static web files.

**Comment by freben** [comment]
Just linking up with the tangentially related #18372 and #18390

**Comment by taras** [comment]
Our client wants teams to release new versions of frontend plugins without rebuilding a container image. They want to be able to allow a team to promote their version of the plugin by choosing a version from a Google Cloud Storage bucket. The `manifest-manager` approach [mentioned by @Rugvip](https://github.com/backstage/backstage/issues/17054#issuecomment-1540025911) seems the best option for us. 

Can we discuss this option at the next Adoption SIG?

**Comment by tumido** [comment]
Hi @freben, @taras. We are working towards a full RFC rewrite based on #18372 and `@backstage/frontend-app-api` package. We want to align this RFC with the new front-end system and offer as seamless integration as possible. So far our experiments look promising and we want to share them as soon as they are in a presentable state. We can briefly discuss this during today's Adoption SIG, however for a more in-depth review I would suggest postponing till the next SIG meeting but no later than that, as we don't want to prolong this indefinitely.

**Comment by efenner-cambia** [comment]
bump

**Comment by phenomuser** [comment]
Hi @tumido , If I may ask, when by when you are planning to have a support of dynamic apps in backstage 
Can we also contribute to this, as we are waiting on this for our use case
Sanga

**Comment by tumido** [comment]
Hi @phenomuser we've demoed our POC during the last Adoption SIG and we're organizing ourselves to re-start a technical discussion around this topic in the [`declarative-intergation`](https://discord.com/channels/687207715902193673/1101159492298743808) channel on Discord.

We have an opinionated implementation ready at https://github.com/janus-idp/backstage-showcase

(see https://github.com/janus-idp/backstage-showcase/blob/main/showcase-docs/dynamic-plugins.md)

however, this implementation is based on the legacy frontend system, not the new `declarative-integration` aka https://github.com/backstage/backstage/issues/18372

Once we sync up on some technical details, we will update this RFC to be based on top of the new frontend system (so we do not end up with 2 declarative frontend systems in Backstage).

**Comment by tumido** [comment]
Closing in favor of [BEP-0002](https://github.com/backstage/backstage/tree/master/beps/0002-dynamic-frontend-plugins). BEP will be a better process for this RFC. 🙂
